beacon-chain/sync/gossipsub_base.go

```diff
@@ -0,0 +1,186 @@
+			defer cancel()
+
+			ctx, span := trace.StartSpan(ctx, "sync.pubsub")
+			defer span.End()
+
+			defer func() {
+				if r := recover(); r != nil {
+					tracing.AnnotateError(span, fmt.Errorf("panic occurred: %v", r))
+					log.WithField("error", r).
+						WithField("recoveredAt", "subscribeWithBase").
+						WithField("stack", string(debug.Stack())).
+						Error("Panic occurred")
+				}
+			}()
+
+			span.SetAttributes(trace.StringAttribute("topic", topic))
+
+			if msg.ValidatorData == nil {
+				log.Error("Received nil message on pubsub")
+				messageFailedProcessingCounter.WithLabelValues(topic).Inc()
+				return
+			}
+
+			if err := b.handler(ctx, msg.ValidatorData.(proto.Message)); err != nil {
+				tracing.AnnotateError(span, err)
+				log.WithError(err).Error("Could not handle p2p pubsub")
+				messageFailedProcessingCounter.WithLabelValues(topic).Inc()
+				return
+			}
+		}
+
+		// The main message loop for receiving incoming messages from this subscription.
+		messageLoop := func() {
+			for {
+				msg, err := sub.Next(s.ctx)
+				if err != nil {
+					// This should only happen when the context is cancelled or subscription is cancelled.
+					if !errors.Is(err, pubsub.ErrSubscriptionCancelled) { // Only log a warning on unexpected errors.
+						log.WithError(err).Warn("Subscription next failed")
+					}
+					// Cancel subscription in the event of an error, as we are
+					// now exiting topic event loop.
+					sub.Cancel()
+					return
+				}
+
+				if msg.ReceivedFrom == s.cfg.p2p.PeerID() {
+					continue
+				}
+
+				go pipeline(msg)
+			}
+		}
+
+		go messageLoop()
+		log.WithField("topic", topic).Info("Subscribed to")
+		b.subscriptions[topic] = sub
+		s.subHandler.addTopic(topic, sub)
+	}
+}
+
+func (b *baseGossipsubTopicFamily) unsubscribeAll() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	for topic, sub := range b.subscriptions {
+		b.cleanupSubscription(topic, sub)
+	}
+
+	b.subscriptions = make(map[string]*pubsub.Subscription)
+}
+
+func (b *baseGossipsubTopicFamily) removeUnwantedTopics(wantedTopics []string) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	topics := wantedTopics
+	neededMap := make(map[string]bool, len(topics))
+	for _, t := range topics {
+		neededMap[t] = true
+	}
+
+	for topic, sub := range b.subscriptions {
+		if !neededMap[topic] {
+			b.cleanupSubscription(topic, sub)
+		}
+	}
+}
+
+func (b *baseGossipsubTopicFamily) cleanupSubscription(topic string, sub *pubsub.Subscription) {
+	s := b.syncService
+	log.WithField("topic", topic).Info("Unsubscribed from")
+	if err := s.cfg.p2p.PubSub().UnregisterTopicValidator(topic); err != nil {
+		log.WithError(err).Error("Could not unregister topic validator")
+	}
+
+	if sub != nil {
+		sub.Cancel()
+	}
+	if err := s.cfg.p2p.LeaveTopic(topic); err != nil {
+		log.WithError(err).Error("Unable to leave topic")
+	}
+
+	if crawler := s.cfg.p2p.Crawler(); crawler != nil {
+		crawler.RemoveTopic(topic)
+	}
```

Comment:
Why do we allow `Crawler` to respond with a nil value? Isn't there always a crawler?